dao/src/main/resources/db/migration/V72__deduplicatePackagesAndProjects.sql

```diff
@@ -0,0 +1,610 @@
+GROUP BY
+  p.id
+UNION 
+SELECT
+  p.id,
+  '' as authors
+FROM
+  packages p
+where
+  NOT EXISTS (
+  SELECT
+    1
+  FROM
+    packages_authors pa
+  WHERE
+    pa.package_id = p.id
+);
+
+-- A view that assigns each package a has value computed from its properties and relations.
+CREATE VIEW packages_with_hashes AS
+SELECT
+  p.id,
+  encode(sha256(convert_to(concat(
+    p.identifier_id,
+    p.vcs_id,
+    p.vcs_processed_id,
+    p.binary_artifact_id,
+    p.source_artifact_id,
+    p.purl,
+    p.cpe,
+    p.description,
+    p.homepage_url,
+    p.is_metadata_only,
+    p.is_modified,
+    pl.licenses,
+    pa.authors,
+    ppl.processed_license_hash
+  ), 'UTF8')),
+    'hex') AS hash
+FROM
+  packages p
+INNER JOIN packages_with_licenses pl ON p.id = pl.id
+INNER JOIN packages_with_authors pa ON p.id = pa.id
+INNER JOIN packages_with_processed_licenses ppl ON p.id = ppl.id;
+
+-- A temporary table that contains only the IDs of the projects that remain after the deduplication.
+CREATE TABLE deduplicated_packages
+(
+    id bigserial PRIMARY KEY
+);
+
+INSERT INTO deduplicated_packages
+SELECT
+  MIN(pwh.id) id
+FROM
+  packages_with_hashes pwh
+GROUP BY pwh.hash;
+
+-- Helper views for projects.
+
+-- A view with the licenses assigned to a project as a single string.
+CREATE VIEW projects_with_licenses AS
+SELECT
+  p.id,
+  string_agg(dl.name, ',' ORDER BY dl.name) licenses
+FROM
+  projects p
+INNER JOIN projects_declared_licenses pdl ON pdl.project_id = p.id
+INNER JOIN declared_licenses dl ON pdl.declared_license_id = dl.id
+GROUP BY
+  p.id
+UNION
+SELECT
+  p.id,
+  '' as licenses
+FROM
+  projects p
+where
+  NOT EXISTS (
+  SELECT
+    1
+  FROM
+    projects_declared_licenses pdl
+  WHERE
+    pdl.project_id = p.id
+);
+
+-- A view with the processed licenses assigned to a project as a single string.
+CREATE VIEW projects_with_processed_licenses AS
+SELECT
+  p.id,
+  string_agg(pdl2.hash, ',' ORDER BY pdl2.hash) processed_license_hash
+FROM
+  projects p
+INNER JOIN processed_declared_licenses pdl3 ON pdl3.project_id = p.id
+INNER JOIN processed_declared_licenses_with_hashes pdl2 ON pdl3.id = pdl2.id
+GROUP BY
+  p.id
+UNION
+SELECT
+  p.id,
+  '' as processed_license_hash
+FROM
+  projects p
+where
+  NOT EXISTS (
+  SELECT
+    1
+  FROM
+    processed_declared_licenses pdl
+  WHERE
+    pdl.project_id = p.id
+);
+
+-- A view with the authors assigned to a project as a single string.
+CREATE VIEW projects_with_authors AS
+SELECT
+  p.id,
+  string_agg(a."name", ',' order by a."name") authors
+FROM
+  projects p
+INNER JOIN projects_authors pa ON pa.project_id = p.id
+INNER JOIN authors a ON	pa.author_id = a.id
+GROUP BY
+  p.id
+UNION
+SELECT
+  p.id,
+  '' as authors
+FROM
+  projects p
+WHERE
+  NOT EXISTS (
+  SELECT
+    1
+  FROM
+    projects_authors pa
+  WHERE
+    pa.project_id = p.id
+);
+
+-- A view with the scopes assigned to a project as a single string.
+CREATE VIEW projects_with_scopes AS
```

Comment:
Just to double check, the dependency graphs are stored as part of the analyzer result and only loosely associated to the projects via the scope name, so they can be ignored here, right?

Reply:
Yes, also a dependency graph contains all projects in the repository; so it cannot be associated with a single project.